### PR TITLE
Enable stable config cache

### PR DESCRIPTION
**Pull request by Goooler**
https://docs.gradle.org/7.6.1/userguide/configuration_cache.html

### CODE REVIEW

**Comment by Goooler** [comment]
```
67 problems were found storing the configuration cache, 9 of which seem unique.
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: cannot serialize object of type 'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: cannot serialize object of type 'org.gradle.api.internal.tasks.DefaultSourceSet', a subtype of 'org.gradle.api.tasks.SourceSet', as these are not supported with the configuration cache.
Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: cannot serialize object of type 'org.gradle.internal.io.LinePerThreadBufferingOutputStream', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: error writing value of type 'java.io.InputStreamReader'
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[[Ljava.lang.Object;@7867171d, [Ljava.lang.Object;@63146832, [Ljava.lang.Object;@1b5d8f2e, [Ljava.lang.Object;@4a48d5c1, [Ljava.lang.Object;@a57ec6a]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[classpath, site]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[java.lang.Class, java.lang.ClassLoader, java.lang.Compiler, java.lang.InheritableThreadLocal, java.lang.Package, java.lang.Process, java.lang.Runtime, java.lang.RuntimePermission, java.lang.SecurityManager, java.lang.System, java.lang.Thread, java.lang.ThreadGroup, java.lang.ThreadLocal]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[org.apache.velocity.runtime.log.AvalonLogChute, org.apache.velocity.runtime.log.Log4JLogChute, org.apache.velocity.runtime.log.CommonsLogLogChute, org.apache.velocity.runtime.log.ServletLogChute, org.apache.velocity.runtime.log.JdkLogChute]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:spotbugsMain` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
```

**Comment by Goooler** [comment]
https://github.com/diffplug/spotless/actions/runs/4274264364/jobs/7440768996

```
19 problems were found storing the configuration cache.
- Task `:lib-extra:spotbugsMain` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsCleanthat` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsCompatKtLint0Dot46Dot0` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsCompatKtLint0Dot47Dot0` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsCompatKtLint0Dot48Dot0` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsDiktat` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsFlexmark` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsGson` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsJackson` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsKtfmt` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsKtlint` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsMain` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsPalantirJavaFormat` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsScalafmt` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
- Task `:lib:spotbugsSortPom` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
```


https://github.com/diffplug/spotless/actions/runs/4274264364/jobs/7440768895
```
67 problems were found storing the configuration cache, 9 of which seem unique.
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: cannot serialize object of type 'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: cannot serialize object of type 'org.gradle.api.internal.tasks.DefaultSourceSet', a subtype of 'org.gradle.api.tasks.SourceSet', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: cannot serialize object of type 'org.gradle.internal.io.LinePerThreadBufferingOutputStream', a subtype of 'java.io.OutputStream', as these are not supported with the configuration cache.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: error writing value of type 'java.io.InputStreamReader'
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[[Ljava.lang.Object;@22f86738, [Ljava.lang.Object;@d29a528, [Ljava.lang.Object;@5556b417, [Ljava.lang.Object;@125cb412, [Ljava.lang.Object;@737cd0c6]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[classpath, site]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[java.lang.Class, java.lang.ClassLoader, java.lang.Compiler, java.lang.InheritableThreadLocal, java.lang.Package, java.lang.Process, java.lang.Runtime, java.lang.RuntimePermission, java.lang.SecurityManager, java.lang.System, java.lang.Thread, java.lang.ThreadGroup, java.lang.ThreadLocal]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:generateMavenPluginHelpMojoSources` of type `de.benediktritter.maven.plugin.development.task.GenerateHelpMojoSourcesTask`: value '[org.apache.velocity.runtime.log.AvalonLogChute, org.apache.velocity.runtime.log.Log4JLogChute, org.apache.velocity.runtime.log.CommonsLogLogChute, org.apache.velocity.runtime.log.ServletLogChute, org.apache.velocity.runtime.log.JdkLogChute]' failed Java Object Serialization
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:not_yet_implemented:java_serialization
- Task `:plugin-maven:spotbugsMain` of type `com.github.spotbugs.snom.SpotBugsTask`: invocation of 'Task.project' at execution time is unsupported.
  See https://docs.gradle.org/7.6.1/userguide/configuration_cache.html#config_cache:requirements:use_project_during_execution
```

**Comment by nedtwigg** [comment]
Thanks for the PR, if all these warnings go away we can merge this. But I'm quite disappointed that the config cache has become stable. I think the design is fatally flawed (#987), writing a Gradle plugin has become far more difficult for little benefit. It reminds me quite a bit of the Eclipse ecosystem when Eclipse 4 and "p2" were first proposed. P2 and E4 were ambitious systems that fundamentally could not deliver on what was promised, and the good parts of Eclipse eventually rotted to nothing due to the friction of the unmet and unmeetable promises of its "next-gen" components (e.g. https://github.com/gradle/gradle/issues/13510 is not possible to implement, but plugin authors have to jump through onerous hoops anyway).

Gradle doesn't have an imminent competitor the same way that Eclipse did, but the configuration cache serialization requirement is imo a catastrophe for the ecosystem.

**Comment by Goooler** [comment]
Suppressed all errors on CI.

**Comment by nedtwigg** [comment]
Thanks for this, linking to all the issues tracking the configuration cache issues was very helpful.

File: CHANGES.md [hunk 1]
```diff
@@ -10,6 +10,8 @@ This document is intended for Spotless developers.
 We adhere to the [keepachangelog](https://keepachangelog.com/en/1.0.0/) format (starting after version `1.27.0`).
 
 ## [Unreleased]
+### Changes
+* We are now opting in to Gradle's new stable configuration cache. ([#1591](https://github.com/diffplug/spotless/pull/1591))
```
Comment (Goooler) [inline CHANGES.md:14]: This is only a behavior of Spotless self, users still need to enable stable CC flags in their projects if they want. 
And next step, we might need to add `--configuration-cache` flags in tests to test compatibitlies.

Comment (nedtwigg) [reply CHANGES.md:14]: Thanks! This changelog is intended mostly for Spotless developers, so I usually put important changes to the build into it. The gradle and maven plugins have their own changelogs for end-users.

**Pull request merged**